Cython/Utility/CythonFunction.c

```diff
@@ -67,8 +68,8 @@ typedef struct {
     PyObject *func_is_coroutine;
 } __pyx_CyFunctionObject;
 
-#define __Pyx_CyFunction_Check(obj)  __Pyx_TypeCheck(obj, __pyx_CyFunctionType)
-#define __Pyx_IsCyOrPyCFunction(obj)  __Pyx_TypeCheck2(obj, __pyx_CyFunctionType, &PyCFunction_Type)
+#define __Pyx_CyFunction_Check(obj)  __Pyx_TypeCheck(obj, ${modulestateglobal_cname}->__pyx_CyFunctionType)
+#define __Pyx_IsCyOrPyCFunction(obj)  __Pyx_TypeCheck2(obj, ${modulestateglobal_cname}->__pyx_CyFunctionType, &PyCFunction_Type)
```

Comment:
I don't think this option will be the final form here, but it's definitely usable for now.

One thing we'll need to handle in future to support is that in "full" module-state mode, we won't always be able to get the module state (e.g. a `cdef` function has no way of getting the module state). In this case, some things won't be able to work because they just need the module state, but things like empty tuples shouldn't fail because the cached version is just an optimization, and it can easily be recreated.

So the final version will have to be able to cope with that. But I don't think it's worth spending too much time designing around that now.